### PR TITLE
filter dynamo versions that support playlist on startup

### PR DESCRIPTION
**Pull request by BogdanZavu**
### Purpose

When I start Revit I want to see the Playlist button enabled if there is at least one version of Dynamo installed that supports Playlist feature. 
If there are more than 1 version , then I'll see the usual version selector dialog when I press the Playlist button. The dialog will show only Dynamo versions that support Playlist and not all Dynamo versions installed ( as it is the case for Dynamo button ).

### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] Snapshot of UI changes, if any.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)
@mjkkirschner 

### FYIs
@kronz

### CODE REVIEW

**Comment by kronz** [comment]
Related question but might not be addressed in this PR.  What happens if a user starts an unsupported version of Dynamo in a session (Revit with then ONLY use the unsupported version until restart of Revit) and then uses Playlist?

**Comment by BogdanZavu** [comment]
@kronz If by an unsupported version of Dynamo you refer to a previous version of Dynamo like 0.9 which doesn't support Playlist then when 0.9 is chosen the Playlist button becomes disabled for the current Revit session.

**Comment by kronz** [comment]
@BogdanZavu ok, thanks.  Seems sensible

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs [hunk 1]
```diff
@@ -36,9 +36,11 @@ public string VersionString
     [Regeneration(RegenerationOption.Manual)]
     public class VersionLoader : IExternalApplication
     {
-        private static UIControlledApplication uiApplication;
+        private static UIControlledApplication uiControlledApplication;
         private AddInCommandBinding dynamoCommand;
+        private AddInCommandBinding dynamoPlayerCommand;
         internal List<DynamoProduct> Products { get; private set; }
+        internal List<DynamoProduct> PlaylistProducts { get; private set; }
```
Comment (mjkkirschner) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: can you add a comment on this property?

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs [hunk 2]
```diff
@@ -60,15 +62,29 @@ public Result OnStartup(UIControlledApplication application)
...
+            PlaylistProducts = new List<DynamoProduct>();
             foreach (var p in dynamoProducts)
             {
                 var path = VersionLoader.GetDynamoRevitPath(p, revitVersion);
                 if (!File.Exists(path))
                     continue;
 
                 Products.Add(p);
+
+                if (p.VersionInfo.Build >= 1 && p.VersionInfo.Major >= 1)
```
Comment (mjkkirschner) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: shouldn't this be `Major` and `Minor` ? https://msdn.microsoft.com/en-us/library/system.version(v=vs.110).aspx#Anchor_6

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs [hunk 3]
```diff
@@ -60,15 +62,29 @@ public Result OnStartup(UIControlledApplication application)
...
+                {
+                    if (Convert.ToInt64(revitVersion) >= 2017)
```
Comment (mjkkirschner) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: can we store all these numbers in some `constants somewhere`

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs [hunk 4]
```diff
@@ -79,29 +95,52 @@ public Result OnStartup(UIControlledApplication application)
...
 
         void executed(object sender, ExecutedEventArgs e)
         {
-            var product = PromptVersionSelectorDialog();
+            var product = PromptVersionSelectorDialog(Products);
             if(product.HasValue)
                 LaunchDynamoCommand(product.Value, e);
         }
 
+        void executedPlaylist(object sender, ExecutedEventArgs e)
```
Comment (mjkkirschner) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: can you add an explicit private access modifier

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs [hunk 5]
```diff
@@ -112,11 +151,25 @@ void canExecute(object sender, CanExecuteEventArgs e)
...
+            var revitApp = ass.CreateInstance("Dynamo.Applications.DynamoRevitApp");
+            revitApp.GetType().GetMethod("OnStartup").Invoke(revitApp, new object[] { uiControlledApplication });
+
+            return Result.Succeeded;
+        }
+
         /// <summary>
         /// Prompts for version selection task dialog
         /// </summary>
         /// <returns>DynamoProduct to launch or null</returns>
```
Comment (mjkkirschner) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: please update the summary (missing parameter)

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs [hunk 6]
```diff
@@ -174,24 +229,35 @@ private bool LaunchDynamoCommand(DynamoProduct product, ExecutedEventArgs e)
...
+            {
+                //Invoke command
+                string message = string.Empty;
+                result = type.GetMethod("ExecuteDynamoCommand").Invoke(revitApp, new object[] { e.GetJournalData(), uiApp });
+                if ((Result)result != Result.Succeeded)
+                    return false;
+            }
+            else
+            {
+                //search it again since it was re-binded in dependent components
```
Comment (mjkkirschner) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: can you clarify this comment a bit more?

**Comment by mjkkirschner** [comment]
@BogdanZavu a few questions - 
also, I don't think it's easily possible now with the current state of RTF - but I want to investigate if we can write a test to check this functionality in the future. I will checkout what we have existing around startup sequence.

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs (hunk 1 above)
Comment (BogdanZavu) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: ok

File: src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs (hunk 3 above)
Comment (BogdanZavu) [inline src/Legacy/DynamoRevitVersionSelector/VersionSelectorApp.cs]: yes

**Comment by BogdanZavu** [comment]
Regarding RTF , if I'm not mistaken you can't have a test right now that chooses Dynamo interactively ; instead you have to specify at the beginning which Dynamo you want to use . 
I think Revit journaling allows some things to be done in addins ( in this case DynamoRevitVersionSelector ) in order for him to know how to press a button in an addin . So just like we have now a Revit journal that starts Revit and Dynamo we could have a journal that starts Revit , presses Dynamo and chooses a version. 
I need to further check this out if it's really possible , I might be wrong.

**Comment by mjkkirschner** [comment]
@BogdanZavu looks good, can you generate PRs to cherry pick these commits to the 1.2 branches and I'll merge them all at the same time.

**Comment by jnealb** [comment]
Does this need testing?

**Comment by mjkkirschner** [comment]
@jnealb definitely.

**Comment by mjkkirschner** [comment]
@BogdanZavu do you need to make any other changes or are these branches all ready for merge?

**Comment by BogdanZavu** [comment]
@mjkkirschner no other changes are needed ; all branches are ready to merge ; Thank you!

**Pull request merged**